### DISCUSSION TITLE
The todo example doesn't pass the render efficiency test

### DISCUSSION THREAD

**Discussion by tylerlong** [pmndrs/valtio#998] [Bug report]
### Bug Description

Please check the test report: https://github.com/tylerlong/todo-render-efficiency/tree/valtio

I copied the code from here https://github.com/pmndrs/valtio/tree/main/examples/todo

I applied `React.memo to` `TodoRow` to reduce the number or re-renders.

There is still one test case failed.  Please help, thanks.

### Reproduction Link

https://github.com/tylerlong/todo-render-efficiency/tree/valtio

**Comment by dai-shi** [comment]
It's probably because you pass the entire `todo` object to TodoRow. Try something like this:
```jsx
<TodoRow key={todo.id} todoId={todo.id} />
```

**Reply by tylerlong** [reply]
Code was copied from here https://github.com/pmndrs/valtio/blob/main/examples/todo/src/App.tsx#L46

I can try to pass an id instead. But how do I access the todo?  Maybe like this:

```tsx
const todos = useTodos();
const todo = todos.find(t => t.id === id)!;
```

It will cause performance issue if there are lots of todos.

**Reply by tylerlong** [reply]
I confirm that passing the id could resolve the render efficiency issue. However, I believe it will cause another kind of performance issue.
Let's say there are `n` todos, to render a list of them the time complexity would be `n ** 2` because we need to run `todos.find` for each todo.

**Reply by dai-shi** [reply]
Maybe passing `index` instead of `id` would make more sense because you can do `todos[index]`.

**Reply by tylerlong** [reply]
Passing `index` will cause other render efficiency issues. For example, removing a todo may cause other todos to re-render because their indexes changed. 
We could tell `React.memo` to ignore the `index` prop. But I am not sure that regular developers would do that.
I hope this library could address this render efficiency issue instead. So that no special code is required to achieve render efficiency.

**Reply by dai-shi** [reply]
I thought about it years ago, and I don't think it's possible for a library to do anything. We have a similar case with Jotai too.

> It will cause performance issue if there are lots of todos.

If `.find` causes a real performance issue, you would need a map, like `todoMap[id]`.

**Reply by tylerlong** [reply]
According to this [test result](https://github.com/tylerlong/todo-render-efficiency?tab=readme-ov-file#testing-results-ordered-by-number-of-stars-on-github), Mobx, Jotai and Manate don't have this issue. Zustand and Valtio have this issue.

**Reply by dai-shi** [reply]
Ah, yeah, Jotai can pass the atom as reference. (I was confused with another case (= omitting React.memo))
To do the same, actually, you can pass todo proxy not snapshot.

```js
function useTodos() {
  const snapShot = useSnapshot(store)

  switch (snapShot.filter) {
    case 'all':
      return [snapShot.todos, store.todos]
    case 'completed':
      return [snapShot.todos.filter((todo) => todo.completed), store.todos.filter((todo) => todo.completed)]
    default:
      throw Error('Error: un supported filter')
  }
}
```

```jsx
function TodoList() {
  console.log("TodoList render");
  const [todos, storeTodos] = useTodos();
  return (
    <div>
      {todos.map((todo, index) => (
        <TodoRow key={todo.id} todo={storeTodos[index]} />
      ))}
    </div>
  );
}
```
